src/test/java/ch/uzh/ifi/hase/soprafs22/service/UserServiceTest.java

```diff
@@ -165,9 +263,413 @@ void updateUser_Conflict(){ //Test what happens when user tries to change userna
+        UserBlackCards ubc = testUser.getUserBlackCards();
+        assertArrayEquals(new List[]{cards}, new List[]{ubc.getBlackCards()});
+    }
+
+    @Test
+    void assignBlackCardsToUser_successMultipleCards() {
+        Mockito.when(userRepository.findById(1L)).thenReturn(testUser);
+        List<BlackCard> cards = new ArrayList<>();
+        cards.add(testBlackCard);
+        BlackCard secondBlackCard = new BlackCard();
+        cards.add(secondBlackCard);
+        userService.assignBlackCardsToUser(testUser.getId(), cards);
+        UserBlackCards ubc = testUser.getUserBlackCards();
+        assertArrayEquals(new List[]{cards}, new List[]{ubc.getBlackCards()});
+    }
+
+    @Test
+    void checkBlackCard_success_givenBlackCard() {
+        Long id = testUser.getId();
+        Mockito.when(userRepository.findById(1L)).thenReturn(testUser);
+        testUser.setUserBlackCards(userBlackCards);
+        assertDoesNotThrow(() -> userService.checkBlackCard(id, testBlackCard));
+    }
+    @Test
+    void checkBlackCard_success_givenNoBlackCard() {
+        Long id = testUser.getId();
+        Mockito.when(userRepository.findById(1L)).thenReturn(testUser);
+        testUser.setUserBlackCards(null);
+        ResponseStatusException e = assertThrows(ResponseStatusException.class, () -> userService.checkBlackCard(id, testBlackCard));
+        assertEquals(HttpStatus.FORBIDDEN, e.getStatus());
+        UserBlackCards ubc = new UserBlackCards();
+        ubc.setBlackCards(null);
+        testUser.setUserBlackCards(ubc);
+        e = assertThrows(ResponseStatusException.class, () -> userService.checkBlackCard(id, testBlackCard));
+        assertEquals(HttpStatus.FORBIDDEN, e.getStatus());
+        List<BlackCard> cards = new ArrayList<>();
+        ubc.setBlackCards(cards);
+        testUser.setUserBlackCards(ubc);
+        e = assertThrows(ResponseStatusException.class, () -> userService.checkBlackCard(id, testBlackCard));
+        assertEquals(HttpStatus.FORBIDDEN, e.getStatus());
+    }
+    @Test
+    void checkBlackCard_success_givenWrongBlackCard() {
+        Long id = testUser.getId();
+        Mockito.when(userRepository.findById(1L)).thenReturn(testUser);
+        testUser.setUserBlackCards(userBlackCards);
+        ResponseStatusException e = assertThrows(ResponseStatusException.class, () -> userService.checkBlackCard(id, otherBlackCard));
+        assertEquals(HttpStatus.FORBIDDEN, e.getStatus());
+    }
+
+    @Test
+    void hasUserAlreadyPlayInGame_true() {
+        // case single Play
+        Play play1 = new Play();
+        play1.setUserId(testUser.getId());
+        testGame.enqueuePlay(play1);
+        assertTrue(userService.hasUserAlreadyPlayInGame(testGame, play1));
+        // case multiple Plays
+        Play play2 = new Play();
+        play2.setUserId(66L);
+        testGame.enqueuePlay(play2);
+        testGame.setUserId(66L);
+        assertTrue(userService.hasUserAlreadyPlayInGame(testGame, play2));
```

Comment:
Please make sure that the game never belongs to the user who puts a play into it. (In the gamecontroller, this is checked on line 235: 
![image](https://user-images.githubusercontent.com/75019468/165055014-f75c1244-73ba-4b47-8a39-4549ccd3f9e3.png)
Also, I don't see why you reset the userId of the testgame (is it not enough to see if user from play1 still has a play in the game?)

Reply:
Please make sure that the game never belongs to the user who puts a play into it. (In the gamecontroller, this is checked on line 235: -- as it is a unit test I understand your concern but I don't think it is an issue in this case..

Also, I don't see why you reset the userId of the testgame (is it not enough to see if user from play1 still has a play in the game?)  -- to check if it also works if it is the second one to check is the one returning true. It could be solved otherwise but this way there is less code..

Reply:
> as it is a unit test I understand your concern but I don't think it is an issue in this case..


Yes, but we might want to decide to add an assertion in enqueue that the play doesn't belong to the game holder, and then the test would fail (I see that it works now, but in the future the game could enforce itself (instead of the controller) that the play doesn't belong to that user - which in my opinion, would make total sense.

> to check if it also works if it is the second one to check is the one returning true. It could be solved otherwise but this way there is less code


But wouldn't assertTrue(userService.hasUserAlreadyPlayInGame(testGame, play1)) do that sufficiently? I'm not sure what we are testing here.

Reply:
has been resolved trough a call and last commit